File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/reader/Telemetry.scala
package org.sunbird.dp.reader

import java.io.Serializable
import java.text.MessageFormat.format
import java.text.SimpleDateFormat
import java.util
import java.util.Date
import java.util.logging.Logger



@SerialVersionUID(8132821816689744470L)
class Telemetry(var map: util.Map[String, Any]) extends Serializable {
 // private val logger = new Logger(this.getClass)

  def add(keyPath: String, value: Any): Boolean = {
    try {
      val lastParent = lastParentMap(map, keyPath)
      lastParent.addChild(value)
      return true
    } catch {
      case e: Exception =>
        //logger.error("", format("Couldn't add value:{0} at  key: {0} for event:{1}", value, keyPath, map), e)
    }
    false
  }

  def getMap: util.Map[String, Any] = map

  def read[T <: Any](keyPath: String): NullableValue[T] = try {
    val parentMap = lastParentMap(map, keyPath)
    new NullableValue[T](parentMap.readChild.get.asInstanceOf[T])
  } catch {
    case e: Exception =>
      //logger.error("", format("Couldn't get key: {0} from event:{1}", keyPath, map), e)
      new NullableValue[T](null.asInstanceOf[T])
  }

  def readOrDefault[T](keyPath: String, defaultValue: T): NullableValue[T] = if (read(keyPath).isNull) new NullableValue[T](defaultValue)
  else read(keyPath).asInstanceOf[NullableValue[T]]

  @throws[TelemetryReaderException]
  def mustReadValue[T](keyPath: String): T = {
    val readValue = read(keyPath)
    if (readValue.isNull) {
      val eid = read("eid")
      var message = keyPath + " is not available in the event"
      if (!eid.isNull) message = keyPath + " is not available in " + eid.value
      throw new TelemetryReaderException(message)
    }
    readValue.value
  }

  private def lastParentMap(map: util.Map[String, Any], keyPath: String) = {
    var parent = map
    val keys = keyPath.split("\\.")
    val lastIndex = keys.length - 1
    if (keys.length > 1) {
      var i = 0
      while ( {
        i < lastIndex && parent != null
      }) {
        var o = null
        if (parent.isInstanceOf[util.Map[_, _]]) o = new ParentMap(parent, keys(i)).readChild.get
        else if (parent.isInstanceOf[util.List[_]]) o = new ParentListOfMap(parent.asInstanceOf[util.List[util.Map[String, Any]]], keys(i)).readChild.get
        else o = new NullParent(parent, keys(i)).readChild.get
        parent = o
        i += 1
      }
    }
    val lastKeyInPath = keys(lastIndex)
    if (parent.isInstanceOf[util.Map[_, _]]) new ParentMap(parent, lastKeyInPath)
    else if (parent.isInstanceOf[util.List[_]]) new ParentListOfMap(parent.asInstanceOf[util.List[util.Map[String, Any]]], lastKeyInPath)
    else new NullParent(parent, lastKeyInPath)
  }

  override def toString: String = "Telemetry{" + "map=" + map + '}'

  override def equals(o: Any): Boolean = {
    if (this.equals(o)) return true
    if (o == null || (this.getClass ne o.getClass)) return false
    val telemetry = o.asInstanceOf[Telemetry]
    if (map != null) map == telemetry.map
    else telemetry.map == null
  }

  override def hashCode: Int = if (map != null) map.hashCode
  else 0

  def id: String = this.read[String]("metadata.checksum").value

  def addFieldIfAbsent[T](fieldName: String, value: T): Unit = {
    if (read(fieldName).isNull) add(fieldName, value.asInstanceOf[AnyRef])
  }

  @throws[TelemetryReaderException]
  def getEts: Long = {
    val ets = this.mustReadValue[Double]("ets")
    ets.toLong
  }

  def getAtTimestamp: String = {
    val timestamp = this.read("@timestamp").value
    val simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
    if (timestamp.isInstanceOf[Number]) {
      val date = new Date(timestamp.asInstanceOf[Number].longValue)
      simpleDateFormat.format(date)
    }
    else this.read[String]("@timestamp").value
  }

  def getSyncts: String = {
    val timestamp = this.read("syncts").value
    val simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
    if (timestamp.isInstanceOf[Number]) {
      val date = new Date(timestamp.asInstanceOf[Number].longValue)
      simpleDateFormat.format(date)
    }
    else this.read[String]("syncts").value
  }
}

File: data-pipeline-flink/dp-telemetry-reader/src/main/java/org/sunbird/dp/reader/Telemetry.java
package org.sunbird.dp.reader;


// import org.ekstep.ep.samza.core.Logger;

import org.sunbird.dp.util.Logger;

import java.io.Serializable;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Map;

import static java.text.MessageFormat.format;

public class Telemetry implements Serializable {

    private static final long serialVersionUID = 8132821816689744470L;

    private Map<String, Object> map;
    private Logger logger = new Logger(this.getClass());

    public Telemetry(Map<String, Object> map) {
        this.map = map;
    }

    public boolean add(String keyPath, Object value) {
        try {
            ParentType lastParent = lastParentMap(map, keyPath);
            lastParent.addChild(value);
            return true;
        } catch (Exception e) {
            logger.error("", format("Couldn't add value:{0} at  key: {0} for event:{1}", value, keyPath, map), e);
        }
        return false;
    }

    public Map<String, Object> getMap() {
        return map;
    }

    public <T> NullableValue<T> read(String keyPath) {
        try {
            ParentType parentMap = lastParentMap(map, keyPath);
            return new NullableValue<>(parentMap.readChild());
        } catch (Exception e) {
            logger.error("", format("Couldn't get key: {0} from event:{1}", keyPath, map), e);
            return new NullableValue<T>(null);
        }
    }

    public <T> NullableValue<T> readOrDefault(String keyPath, T defaultValue) {
        return read(keyPath).isNull() ? new NullableValue<>(defaultValue) : read(keyPath);
    }

    public <T> T mustReadValue(String keyPath) throws TelemetryReaderException {
        NullableValue<T> val = read(keyPath);
        if (val.isNull()) {
            NullableValue<String> eid = read("eid");
            String message = keyPath + " is not available in the event";
            if (!eid.isNull()) {
                message = keyPath + " is not available in " + eid.value();
            }
            throw new TelemetryReaderException(message);
        }

        return val.value();
    }

    private ParentType lastParentMap(Map<String, Object> map, String keyPath) {
        Object parent = map;
        String[] keys = keyPath.split("\\.");
        int lastIndex = keys.length - 1;
        if (keys.length > 1) {
            for (int i = 0; i < lastIndex && parent != null; i++) {
                Object o;
                // System.out.println("Keypath = " + keyPath);
                // System.out.println("Parent instance = " + parent.getClass());
                if (parent instanceof Map) {
                    o = new ParentMap((Map<String, Object>) parent, keys[i]).readChild();
                } else if (parent instanceof List) {
                    o = new ParentListOfMap((List<Map<String, Object>>) parent, keys[i]).readChild();
                } else {
                    o = new NullParent(parent, keys[i]).readChild();
                }
                parent = o;
            }
        }
        String lastKeyInPath = keys[lastIndex];
        if (parent instanceof Map) {
            return new ParentMap((Map<String, Object>) parent, lastKeyInPath);
        } else if (parent instanceof List) {
            return new ParentListOfMap((List<Map<String, Object>>) parent, lastKeyInPath);
        } else {
            return new NullParent(parent, lastKeyInPath);
        }
    }

    @Override
    public String toString() {
        return "Telemetry{" + "map=" + map + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || this.getClass() != o.getClass()) return false;

        Telemetry telemetry = (Telemetry) o;

        return map != null ? map.equals(telemetry.map) : telemetry.map == null;
    }

    @Override
    public int hashCode() {
        return map != null ? map.hashCode() : 0;
    }

    public String id() {
        return this.<String>read("metadata.checksum").value();
    }

    public <T> void addFieldIfAbsent(String fieldName, T value) {
        if (read(fieldName).isNull()) {
            add(fieldName, value);
        }
    }

    public long getEts() throws TelemetryReaderException {
        double ets = this.<Double>mustReadValue("ets");
        return (long) ets;
    }

    public String getAtTimestamp(){
        Object timestamp = this.read("@timestamp").value();
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
        if ( timestamp instanceof Number){
            Date date = new Date(((Number) timestamp).longValue());
            return simpleDateFormat.format(date);
        } else {
            return this.<String>read("@timestamp").value();
        }
    }
    
    public String getSyncts(){
        Object timestamp = this.read("syncts").value();
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
        if ( timestamp instanceof Number){
            Date date = new Date(((Number) timestamp).longValue());
            return simpleDateFormat.format(date);
        } else {
            return this.<String>read("syncts").value();
        }
    }

}
